### PR TITLE
[Clang][Sema] Allow access to a public template alias declaration that refers to friend's private nested type

### LINKED ISSUE #25708: Access to a public template alias declaration that refers to friend's private nested type is not allowed.
|  |  |
| --- | --- |
| Bugzilla Link | [25334](https://llvm.org/bz25334) |
| Version | 3.7 |
| OS | Linux |
| Attachments | [test code (also in the text)](https://user-images.githubusercontent.com/92601579/143752318-71f7ca20-13c3-4f8b-aef1-a802cddd9b53.gz) |
| CC | @DougGregor,@ecatmur |

## Extended Description 

The test case that fails is:

```cpp
struct FooAccessor
{
    template <typename T>
    using Foo = typename T::Foo;
};

class Type
{
    friend struct FooAccessor;
    
    using Foo = int;
};

int main()
{
    FooAccessor::Foo<Type> t;
}
```

The result is:

```console
$ clang++ test.cpp -std=c++11
test.cpp:4:5: error: 'Foo' is a private member of 'Type'
    using Foo = typename T::Foo;
    ^
test.cpp:11:11: note: implicitly declared private here
    using Foo = int;
          ^
1 error generated.
```

### PR DESCRIPTION
**Pull request by jcsxky**
This patch attempts to fix https://github.com/llvm/llvm-project/issues/25708
Current access check missed qualifier(`NestedNameSpecifier`) in friend class checking. Add it to `Records` of `EffectiveContext` by changing the `DeclContext` makes `MatchesFriend` work.

### CODE REVIEW

**Comment by jcsxky** [comment]
gently ping~

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -409,7 +413,7 @@ RISC-V Support
 CUDA/HIP Language Changes
 ^^^^^^^^^^^^^^^^^^^^^^^^^
 
-- PTX is no longer included by default when compiling for CUDA. Using 
+- PTX is no longer included by default when compiling for CUDA. Using
```
Comment (erichkeane) [inline clang/docs/ReleaseNotes.rst:416]: Unrelated change, please revert.

**Review by erichkeane — APPROVED** [review]
1 nit, else LGTM.

File: clang/lib/Sema/SemaTemplate.cpp [hunk 2]
```diff
@@ -4342,10 +4342,17 @@ QualType Sema::CheckTemplateIdType(TemplateName Name,
     InstantiatingTemplate Inst(*this, TemplateLoc, Template);
     if (Inst.isInvalid())
       return QualType();
+    if (!AliasTemplate->getDeclContext()->isFileContext()) {
+      ContextRAII SavedContext(*this, AliasTemplate->getDeclContext());
```
Comment (zyn0217) [inline clang/lib/Sema/SemaTemplate.cpp:4346]: nit: This can be simplified using `std::optional`:
```cpp
std::optional<ContextRAII> C;
if (!A.getDeclContext()->isFileContext())
  C.emplace(*this, A->getDeclContext());
```

Comment (erichkeane) [reply clang/lib/Sema/SemaTemplate.cpp:4346]: > nit: This can be simplified using `std::optional`:
> 
> ```c++
> std::optional<ContextRAII> C;
> if (!A.getDeclContext()->isFileContext())
>   C.emplace(*this, A->getDeclContext());
> ```

Neat trick, I like it!

Comment (jcsxky) [reply clang/lib/Sema/SemaTemplate.cpp]: Looks more clear. Thanks for your guidance!

File: clang/docs/ReleaseNotes.rst (hunk 1 above)
Comment (jcsxky) [reply clang/docs/ReleaseNotes.rst]: Fixed.

File: clang/test/SemaTemplate/PR25708.cpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
+// RUN: %clang_cc1 -std=c++11 -verify %s
+// RUN: %clang_cc1 -std=c++14 -verify %s
+// RUN: %clang_cc1 -std=c++17 -verify %s
+// RUN: %clang_cc1 -std=c++20 -verify %s
```
[the remark this answers is not in the document]
Comment (jcsxky) [reply clang/test/SemaTemplate/PR25708.cpp]: Fixed.

File: clang/docs/ReleaseNotes.rst [hunk 4]
```diff
@@ -259,6 +259,10 @@ Bug Fixes in This Version
   operator.
   Fixes (#GH83267).
 
+- Allow access to a public template alias declaration that refers to friend's
```
[the remark this answers is not in the document]
Comment (jcsxky) [reply clang/docs/ReleaseNotes.rst]: Updated and put it to the right place.
